WIP: Ubuntu kata guest selinux rfc

alextibbles opened the pull request:
### Elements of this PR
- Changes to osbuilder to build Ubuntu images which have SELinux enforcement support enabled, including policy module coverage of Kata customizations of the guest OS, the kata-agent and minimal workload function.
- A change to runtime to enable sending an SELinux label to the agent when SELinux is not active on the host.
Question: should these be split into 2 PRs?

### Progress
- [x] tools/osbuilder/rootfs-builder: Enable building of Ubuntu guest images with SELinux enabled, with the necessary packages included.
- [x] tools/osbuilder/image-builder: Allow controlling whether SELinux is enforced, to enable debugging.
- [x] runtime: Allow launching an SELinux-enabled guest without host SELinux, passing a label to be applied to the workload container inside the guest.
- [x] Adapt the distro-provided SELinux policy to the kata guest image contents.
- [x] Understand and manage the size of the resulting image file.
- [x]  Ensure all Ubuntu-specific code is conditionalized.
- [x]  Document the SELinux policy allow rule origins.
- [ ]  Address Copilot review comments.
- [ ] Implement workload mount labeling in some fashion and remove unlabeled rules.

Work in progress for feedback.

### Limitations of current work
- Currently there is no change to runtime for transmitting mount label instructions to the agent. At least in my testing, this results in the workload mounts being unlabeled. The workload SELinux rules therefore give access to unlabeled files to the workload container_t type, which is not fail-secure in the event of other failures to label files. #11789
- One potentially important workflow for using this would be the various stages the life-cycle of workload-specific policies. There are several ways this might be supported that doesn't involve directly modifying the osbuilder inputs, generating custom images and inspection of the results, within the constraints of read-only guest rootfs. For example, it might be nice to be able to use a tool like https://github.com/containers/udica to generate a policy that would apply inside the guest, which might need adapting to the kata context.
- One likely important concern for SELinux guest usage is the ability to detect violation of policy. I.e. some way of delivering the AVC logs generated in the guest system components where they can be inspected for security response and debugging. #11778
- The SELinux policy isn't modularized between base OS, kata-agent and workload. Theoretically, some portion of the base OS rules might be eligible for upstream (in Ubuntu). I believe the question of how Kata should distribute kata-specific policies has been discussed for host-side confinement extensively - this looks similar. Workload-specific policy seems largely separate, assuming the packages policy provides reasonable defaults.
- All tools for administering SELinux policy are deployed as part of the guest image, which will be inert since they can't operate on the immutable filesystem. This increases the image size by ~90 MB purely from package installation, and approximately 130MB in final image size. Building the policy files and deploying them into the image from the builder containers could perhaps address this.

fidencio commented on tools/osbuilder/image-builder/image_builder.sh:
```diff
@@ -405,6 +411,17 @@ create_disk() {
+	if [ -f ${mount_dir}/etc/os-release ]; then
+		local DISTRO=`sed -n 's/^NAME=\(.*\)$/\1/p' < ${mount_dir}/etc/os-release`
+		if [ "${DISTRO}" == "\"Ubuntu\"" ]; then
+			chroot "${mount_dir}" /usr/bin/chcon "system_u:system_r:dockerd_t:s0" /usr/bin/kata-agent
```
Hmm. Why `dockerd_t` instead of `container_t`?

fidencio replied:
Nevermind, the agent should be `dockerd_t`, indeed.

alextibbles replied:
Thank you for taking a look. The fact the question was raised suggests a comment should be added to provide the information directly.

alextibbles commented:
Many thanks @fidencio for taking another look. Checking my notes, where I believe I got stuck was mount labeling not working how I would like it to (acting as a fall back adding into a labeled tree), for joining the build-time labeling to runtime for the workload. Perhaps a more extensive integration between the agent and the guest SELinux stack is necessary to get something worthwhile working. However, my experiments could easily have missed something...

alextibbles committed:
runtime: send SELinux label to agent, when host SELinux is not active
tools/osbuilder: enable building Ubuntu guest images with SELinux enabled

Add option for permissive mode SELinux image builds for debugging with SELINUX_ENFORCING=no.
Provide a SELinux module sufficient for a minimal workload.

Fixes: #11739, #11738

Signed-of-by: Alex Tibbles <alex@bleg.org>
